File: ml/check_oracle.py
#!/usr/bin/env python3
"""학습본(PyTorch) 예측값을 덤프 — JS 배포본과 대조하기 위한 정답지.
JS 쪽 1층은 가중치를 [입력,은닉]으로 전치해 희소 누적하므로, 전치/인덱싱 실수가
조용히 틀린 봇을 만들 수 있다. 그 사고를 수치로 잡는다.
사용: .venv/bin/python check_oracle.py <oracle.pt> <data.bin> <out.json> [n=200]
"""
import json, sys
import numpy as np
import torch
from train_oracle import Oracle, S_ORC, REC

pt, binf, outf = sys.argv[1], sys.argv[2], sys.argv[3]
n = int(sys.argv[4]) if len(sys.argv) > 4 else 200

sd = torch.load(pt, map_location="cpu")
net = Oracle(sd["o.0.bias"].shape[0], sd["o.2.bias"].shape[0])
net.load_state_dict(sd)
net.eval()

m = np.memmap(binf, dtype=np.float32, mode="r")
rows = m[: (len(m) // REC) * REC].reshape(-1, REC)[:n]
x = torch.from_numpy(np.asarray(rows[:, :S_ORC]))
with torch.no_grad():
    y = net(x).numpy()

json.dump({"n": int(n), "pred": [float(v) for v in y]}, open(outf, "w"))
print(f"덤프 {n}건 -> {outf}  (예측 범위 {y.min():+.3f} ~ {y.max():+.3f})")

File: ml/check-oracle.js
#!/usr/bin/env node
/* 배포본(oracle-value.js) 예측을 학습본 덤프와 대조.
 * 사용: node ml/check-oracle.js <oracle.json> <data.bin> <py-pred.json>
 */
'use strict';
var fs = require('fs'), path = require('path');
var OR = require(path.join(__dirname, '..', 'shared', 'oracle-value.js'));

var oracle = OR.load(process.argv[2]);
var buf = fs.readFileSync(process.argv[3]);
var f = new Float32Array(buf.buffer, buf.byteOffset, Math.floor(buf.byteLength / 4));
var REC = OR.S_ORC + 1;
var py = JSON.parse(fs.readFileSync(process.argv[4], 'utf8'));

var maxDiff = 0, worst = -1, row = new Float32Array(OR.S_ORC);
for (var i = 0; i < py.n; i++) {
  for (var j = 0; j < OR.S_ORC; j++) row[j] = f[i * REC + j];
  var d = Math.abs(oracle.valueDense(row) - py.pred[i]);
  if (d > maxDiff) { maxDiff = d; worst = i; }
}
console.log('대조 ' + py.n + '건 · 최대 절대오차 ' + maxDiff.toExponential(2) + ' (행 ' + worst + ')');
if (maxDiff > 1e-3) { console.log('❌ 불일치 — 전치/인덱싱 확인 필요'); process.exit(1); }
console.log('✅ 학습본과 배포본 일치 (float32 오차 범위)');
